Interactive dialogs

lumayaa opened the pull request:
Add Interactive Menu Navigation to lmd_terminal and Extend lmd_playercheck for Profession Locking
This PR introduces interactive menu functionality to lmd_terminal entities and extends lmd_playercheck to support side-specific access control.

Terminal Interaction (Spawnflags 4)
When spawnflags is set to 4, lmd_terminal becomes interactive:

Navigate between choices using Move Forward and Move Backward.

Confirm a highlighted choice using the Use key.

The player becomes invulnerable while interacting with the terminal.

New Keys:

- Color1: Color code for selection brackets.
- Color2: Color code for the "Cancel" button.
- selectsnd: Sound played when confirming a choice.
- navsnd: Sound played when navigating between options.
- cancelsnd: Sound played when cancelling the menu.

Additional Feature: Early Version of lmd_trainer
This PR also includes the groundwork for a new feature called lmd_trainer. It allows players to upgrade or downgrade their skills via an interactive interface. While functional, this feature is still in development and will be refined in future updates.

lmd_playercheck Extension
The lmd_playercheck entity now supports a new key:

sideProf: Restricts access to either Jedi or Sith players.

- sideProf,1 = Jedi only
- sideProf,2 = Sith only

lumayaa committed:
Interactive Lmd_terminals.

Interactive Lmd_terminals.

lumayaa committed:
Added Animation while using Interactive terminal.

lumayaa committed:
Overhaul InteractiveMenus.

Fixed a Bug where the animation when interacting with a terminal wasn't properly looping.
Fixed a Bug where you could open your saber while interacting with a terminal.
Fixed a Bug where you could jump while interacting with a terminal.
Added key to lmd_terminal: color1 = used for the brackets and cancel highlight color.
Added key to lmd_terminal: color2 = used for the command number color.
Added key to lmd_terminal: selectsnd = default = "sound/movers/switches/switch1.mp3". - The sound playing when confirming a choice.
Added key to lmd_terminal: navsnd = default = "sound/interface/menuroam.mp3". - The sound playing when navigating the lmd_menu.
Added key to lmd_terminal: cancelsnd = default = "sound/interface/esc.mp3". - The sound playing when canceling the lmd_menu.
Added numbers to the choices and wrap the selection brackets around them instead of the entire choice.
Fixed a Bug where the Menu would be displayed a few seconds after confirming or aborting a lmd_menu.
Made the player invulnerable while interacting with lmd_menu.
Improved menu stability: the layout no longer shifts when moving between options.

lumayaa committed:
Fixed a bug where lmd_terminal would print to console after aborting or confirming a choice from lmd_menu.

lumayaa committed:
Adjusted lmd_menu so you can chain them.

In an lmd_terminal you can target another lmd_terminal with spawnflags 4 of course. Example:
/place lmd_terminal * targetname,oldTerminalTarget,cmd,Yeh it's chained.,message,Hello again.

lumayaa committed:
Typewriting effect.

Added 2 keys to lmd_terminal:

- choiceDelay
- messageDelay

if 0 on both: Everything appears instantly.
if choiceDelay > 0: choices appear one after one for that delay set after messge appeared.
if messageDelay > 0: message comes as a typewriting effect with the delay set.

lumayaa committed:
Fixed a bug where we returned on spawnflags 4, before firing usetarget.

lumayaa committed:
Bug fixes.

Made it harder for server command overflow to happen.
Fixed a bug where our hitbox isn't solid while we are interacting.

slash-fr commented on game/Lmd_Entities_Ents.c:
```diff
@@ -1219,1092 +1374,1505 @@ void Think_SpawnNewDoorTrigger( gentity_t *ent );
+        }
+    }
+
+    if (player->client->Lmd.lmdMenu.messageCharsVisible >= (int)strlen(menu->message))
```
Visual Studio points out that:
> ⚠️ C6387: 'menu->message' could be '0':  this does not adhere to the specification for the function 'strlen'.

Can be fixed by checking `menu->message`:
```c
if (menu->message && player->client->Lmd.lmdMenu.messageCharsVisible >= (int)strlen(menu->message))
```

slash-fr commented on game/Lmd_Entities_Ents.c:
```diff
@@ -1219,1092 +1374,1505 @@ void Think_SpawnNewDoorTrigger( gentity_t *ent );
+    }
+
+    if (player->client->Lmd.lmdMenu.messageCharsVisible >= (int)strlen(menu->message) &&
+        player->client->Lmd.lmdMenu.choicesVisible >= menu->count)
```
Should also check `menu->message` here:
```c
    if (menu->message && player->client->Lmd.lmdMenu.messageCharsVisible >= (int)strlen(menu->message) &&
        player->client->Lmd.lmdMenu.choicesVisible >= menu->count)
```

slash-fr commented on game/Lmd_Entities_Ents.c:
```diff
@@ -2376,854 +2944,1262 @@ qboolean CheckRestrictWeaponFire(gentity_t *player) {
-	}
+const entityInfoData_t lmd_trainer_keys[] = {
+    {"color1", "Movement speed.  Default 100."},
+    {"color2", "Damage to inflict when blocked."},
```
Copy pasted descriptions 😇

lumayaa committed:
Lmd_Trainer bug fixing.

Fixed a bug where, the skill menu would instantly reopen.
Doesn't let us use the trainer when we aren't logged in.
Fixed Color descriptions.
Fixed wrong sound files in lmd_trainer.
Added missing nav sound to lmd_trainer.

slash-fr reviewed (approved):
**TODO:** Fix the `lmd_trainer` behavior for **Mercenary** users in a subsequent PR.
